Adds flag statement (TODO, FIXME, ...) parsing for docs

samueleaton opened the pull request:
Adds parsing for flag statements (TODO, FIXME, BUG, ...) for docs as discussed in issue **#3460**

I had to create a workaround because `TODO:` and other keywords are not part of the markdown parser.

The code I added:  
(1) Makes sure that lines starting with `TODO:` are isolated into their own paragraph  
(2) Splits the HTML output by paragraph, and then replaces `TODO:` with the relevant HTML

If there is another way to do this (e.g. with the markdown parser) point me in the right direction and I'll see what I can do.

Screenshot

<img width="1280" alt="screen shot 2016-10-23 at 10 18 17 am" src="https://cloud.githubusercontent.com/assets/9119269/19627693/2dcc5624-990a-11e6-893c-00aea40e0760.png">

BlaXpirit commented:
What is the point of this?

Sija commented:
@BlaXpirit @samueleaton wrote down his reasoning behind this PR in #3460.
@samueleaton I'd add support for `FIXME` too.

ysbaddaden commented:
As well as NOTE and OPTIMIZE.

samueleaton commented:
Should TODO, NOTE, FIXME, and OPTIMIZE be different colors or all orange?

Different Colors: 

![screen shot 2016-11-07 at 6 06 09 pm](https://cloud.githubusercontent.com/assets/9119269/20082686/f92d857a-a514-11e6-8866-cb3609c6c4db.png)

All Orange:

![screen shot 2016-11-07 at 6 07 14 pm](https://cloud.githubusercontent.com/assets/9119269/20082699/0c7d815c-a515-11e6-98f9-2655d82cf601.png)

keplersj commented:
I like the different color! Perhaps also add support for `DEPRECATED`?

samueleaton commented:
@keplersj If we were to do DEPRECATED, would it be the same format with a colon? Would the text that follows the colon be an explanation on why it is deprecated?

Like this:

```ruby
# one of our beloved customers
#
# TODO: create methods for this class
#
class Customer

  # fetches a customer from the database by id
  #
  # DEPRECATED: Use *retrieve* instead
  # TODO: query db for customer  
  # TODO: return nil if customer not found with id  
  # TODO: return an instance of `Customer` if customer exists  
  def self.fetch(customer_id : String)
  end
end
```

I personally think deprecated should be on the same line as the method description:

```ruby
# one of our beloved customers
#
# TODO: create methods for this class
#
class Customer

  # fetches a customer from the database by id (DEPRECATED)
  #
  # TODO: query db for customer  
  # TODO: return nil if customer not found with id  
  # TODO: return an instance of `Customer` if customer exists  
  def self.fetch(customer_id : String)
  end
end
```

This way the deprecated flag would be next to the method definition rather than underneath it. It would also make it so you can't put multiple DEPRECATED flags, which seems appropriate.

keplersj commented:
@samueleaton I like the first one better, I think it's a lot like [Scala's deprecated annotation](http://www.scala-lang.org/api/rc2/scala/deprecated.html).

samueleaton commented:
As of the latest commit, the following will be parsed in the docs:

- Red flags: `FIXME`, `BUG`, `DEPRECATED`  
- Orange flags: `TODO`  
- Yellow flags: `NOTE`  
- Green flags: `OPTIMIZE`

samueleaton commented:
BUG and FIXME are seem similar. Is there one that people are more fond of? I am in favor of BUG. Its shorter and is a more common word (easier to remember).

ysbaddaden commented:
FIXME is more common. It may no be a bug, something poorly written for example.

RX14 commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -241,6 +251,17 @@ class Crystal::Doc::Generator
 
+  # adds extra line break to TODO lines
+  def isolate_flag_lines(string)
+    string.split(/\n/).map do |line|
```
```cr
String.build do |io|
  string.each_line.join('\n', io) do |line|
    if line =~ /^ ?(TODO|FIXME|BUG|NOTE|OPTIMIZE|DEPRECATED):/
      io << line << '\n'
    else
      io << line
    end
  end
end
```

RX14 replied:
I feel like the regex here and the regex on line 235-240 could be replaced by a method which takes a line and returns `{flag, text}` or `nil`. We can even combine them, as HTML passes through markdown unimpeded.

For example:
```cr
if flag, text = parse_flag(line)
  io << %(<span class="flag flag-#{flag.downcase}">#{flag}</span> #{text}\n)
else
  io << line
end
```

which requires only one pass.

samueleaton replied:
Are you saying to remove the regex completely or just to consolidate the regex into a single method?

RX14 replied:
I was assuming that html passed through the markdown parser unescaped, because that's what I've seen most engines do. Github lets you use certain html tags in markdown. 

I think that it is probably a bad *default* to let html through, but it should at least be an option. However if we went this route we would have to sanitize html to include only certain tags (github has a list they let through their markdown). It might be worth looking at nature markdown implementations to see how this is done.

RX14 replied:
However my first suggestion with the each_line and join with block are still valid with this workaround.

samueleaton commented:
@ysbaddaden Two Questions: (1) Do we want to require that every flag ends in a colon? (2) Do we want all flags to be required to be uppercase?

### Ending in Colon

For example, here the DEPRECATED would become a flag:

```ruby
class Customer

  # fetches a customer from the database by id (DEPRECATED)
  #
  # DEPRECATED: This method was buggy and is deprecated 
  def self.fetch(customer_id : String)
  end
end
```

But here the DEPRECATED would not become a flag:

```ruby
class Customer

  # fetches a customer from the database by id (DEPRECATED)
  #
  # DEPRECATED
  def self.fetch(customer_id : String)
  end
end
```

### Require Uppercase

Do we want this to work?

```ruby
class Customer

  # fetches a customer from the database by id (DEPRECATED)
  #
  # deprecated: This method was buggy and is deprecated 
  def self.fetch(customer_id : String)
  end
end
```

samueleaton commented:
@Sija Gotcha. I reverted the commit.

ysbaddaden commented:
No big rule, but I would say that the colon is optional, as long as the definition is on its own line, whereas uppercase is mandatory.

That is:

```
TODO
TODO incredible feature
TODO: support ARM
```

samueleaton commented:
I updated the code to also flag the keywords in the main project README.

**Before**  
![readme-before](https://cloud.githubusercontent.com/assets/9119269/20117741/192111f8-a5bf-11e6-8963-46b57c1f2478.png)

**After**  
![readme-after](https://cloud.githubusercontent.com/assets/9119269/20117746/1d9e5fec-a5bf-11e6-9c57-732088981331.png)

Sija commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -251,7 +243,22 @@ class Crystal::Doc::Generator
+  # Adds extra line break to flag keyword lines
   def isolate_flag_lines(string)
     string.split(/\n/).map do |line|
       if line =~ /^ ?(TODO|FIXME|BUG|NOTE|OPTIMIZE|DEPRECATED):?/
```
I'd use a hash containing flags and corresponding colors — or even just an array with flags, moving the colors into the css itself.

samueleaton replied:
@Sija  Something like this?

```ruby
FLAG_COLORS = {
  "BUG" => "red",
  "DEPRECATED" => "red",
  "FIXME" => "red",
  "NOTE" => "yellow",
  "OPTIMIZE" => "green",
  "TODO" => "orange"
}
FLAGS = FLAG_COLORS.keys

def generate_flags(string)
  FLAGS.reduce(string) do |str, flag|
    flag_regexp = /<p>\s*#{flag}(:| )?/
    element_sub = %(<p><span class="flag #{FLAG_COLORS[flag]}">#{flag}</span> )
    str.gsub(flag_regexp, element_sub)
  end
end
```

Sija replied:
yep, and you could reuse it in `isolate_flag_lines` method too.

samueleaton commented:
Here is the latest on colors:

![flag_colors](https://cloud.githubusercontent.com/assets/9119269/20142815/9425ef60-a653-11e6-9d1c-81b10ea4be5c.png)

luislavena commented:
@samueleaton seems that the failure in Travis is caused by formatting issues. Can you reformat `tools/doc/generator.cr` using the compiler from the checkout (not the 0.19.4 release) and push again?

Thank you.

Sija commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -226,9 +237,11 @@ class Crystal::Doc::Generator
   def doc(context, string)
-    String.build do |io|
-      Markdown.parse string, MarkdownDocRenderer.new(context, io)
+    str = isolate_flag_lines string
```
why create new variable? couldn't that just be `string = isolate_flag_lines string`?

samueleaton replied:
Been doing some functional programming and it is showing 😅 . I can change it.

samueleaton replied:
I could probably just remove that line altogether and do this 

```ruby
Markdown.parse isolate_flag_lines(string), MarkdownDocRenderer.new(context, io)
```

Sija replied:
Keeping it in a separate line seems clearer to me.

Sija commented on src/compiler/crystal/tools/doc/html/css/style.css:
```diff
@@ -349,6 +349,43 @@ code {
+  padding: 2px 4px 1px;
+  border-radius: 3px;
+  margin-right: 3px;
+  font-size: 11px;
```
I'd suggest folding multiple `border` declarations into:
```css
border: 1px solid transparent;
```
and l8r in `span.flag.<color>`:
```css
border-color: #color;
```

samueleaton replied:
This came to mind but I i didn't think it was necessary. In hindsight it is better that way. Good eye. 👁 👍

Sija replied:
:cookie:

RX14 commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -52,7 +63,7 @@ class Crystal::Doc::Generator
     end
 
-    File.write "#{@dir}/index.html", MainTemplate.new(body, types, repository_name)
+    File.write "#{@dir}/index.html", MainTemplate.new(generate_flags(body), types, repository_name)
```
I actually thing that instead of duplicating changes in this method, this can be replaced with a call to `doc(program_type, body)`.

RX14 replied:
This hasn't been fixed.

samueleaton replied:
Fixed it. I even made an optimization where it won't even run the markdown parser on an empty string for the README if the file doesn't exist.

Sija commented:
GTG?

luislavena commented:
I think is good enough to be merged, @samueleaton can you [rebase and squash](https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#maintain-clean-pull-requests) all the changes?

Thank you.

asterite commented:
O still don't quite understand why you wouldn't put a big TODO in public documentation. Is it to encourage pull requests? Is it a remainder for self? Is it to make a library look incomplete?

I'm not against this change, I just can't see the usefulness.

luislavena commented:
Hello @asterite I'm not sure to follow:

> why you wouldn't put a big TODO in public documentation

You mean the opposite?

Since docs of your project (public or private) will be processed by `crystal doc`, I personally find it helpful to highlight any of these notes in our methods or classes.

That of course doesn't cancel or render pointless to have a `TODO` section on the README file or even a `ROADMAP.md` document.

I don't think they are mutually exclusive.

samueleaton commented:
@luislavena @asterite Squashed it.

@asterite You may not always want a big TODO in your public docs, but you might. I can definitely see people wanting to put an an OPTIMIZE, DEPRECATED, or NOTE flag in public docs.

TODO is really good for anyone to see that there is something to work on in the codebase, whether private or public.

Sija commented:
Would be a nice doc enhancement for 0.20 release.

asterite commented:
If many think this is a useful enhancement then let's merge this once it's good to go :-)

samueleaton commented:
@asterite its squashed and the reviews are fixed. You will also need to merge the pending request at [crystal-book #45](https://github.com/crystal-lang/crystal-book/pull/45).

RX14 commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -42,15 +53,11 @@ class Crystal::Doc::Generator
-    body = String.build do |io|
-      Markdown.parse body, MarkdownDocRenderer.new(program_type, io)
-    end
+    body = if filename
```
I much prefer the old style of having `body =` inside the block. Also isn't this `String.build` redundant? Doesn't `doc` return a string anyway?

samueleaton replied:
fixed. although I would have done...

```
body = filename ? doc(program_type, File.read(filename)) : ""
```

... but i don't think it would have been received well

Sija commented:
@asterite :shipit:

asterite commented:
Is this ready to merge? If so, I'll merge it

Sija commented:
Seems like it :)

RX14 commented:
I can't find anything to complain about any more, at least :)

Maybe if we hit performance issues in the doc generator the regex can be replaced but I think it's a non-issue for now.

asterite commented:
I just tried it and the resulting doc is broken. There are double new lines everywhere. See screenshot:

<img width="664" alt="screenshot 2016-11-19 21 21 54" src="https://cloud.githubusercontent.com/assets/209371/20459387/510555ba-ae9e-11e6-85bf-c281407501cb.png">

So for now I won't merge this.

samueleaton commented:
@asterite I found and fixed the problem. I expected `string.each_line.join("", io) do |line, io|` to remove the `\n` at the end of each `line` in the block. (as would happen with `string.split("\n").each do |line|`).

asterite commented:
Looks good now, thank you! ❤️

Sija commented:
Flag statements split into more than 1 line end up like this (which probably is a bug):

![image](https://cloud.githubusercontent.com/assets/988/20822923/6414c8d8-b84f-11e6-94f6-474a76fee48d.png)

RX14 commented:
@Sija I think that that is correct. Your example looks like 2 lines to me. However there should be a way to have multi line todos, maybe indent?

Sija commented:
@RX14 I think that particular paragraphs and code blocks are supposed to be surrounded by an empty line, which should allow you for writing multiline statements without additional indentation.

```cr
# Multiline comments have all lines
# concatenated until empty line is found.
#
# This would be next paragraph.
```

RX14 commented:
So what's the difference between these?
```cr
# TODO this is a long string which is a
# TODO which wraps into another line.
```

```cr
# TODO this is todo 1
# TODO this is todo 2
```

I meant to first one to render "this is a long string which is a TODO which wraps into another line." and the second to render 2 TODOs.

samueleaton commented:
@Sija That is how it was intended. I agree with @RX14 that there should be support for multiline flags.

However, I would think that users are more likely to have 2 or more flags in a row than have a really long flag, so I don't think a set of flags should have an extra blank line between each flag.

Sija commented:
I agree that `TODO` flags are rather short, but `NOTE` ones easily could span far more than 80 characters which is often used as a soft wrap limit. I also agree that every flag separated by an extra
newline would be lengthy. Indentation seems like a good way to handle that.

samueleaton commented:
@sija I'm working on a project and now agree with you about the `NOTE`s extending more than a single line. 😭 

So I think I'll make another PR with the following change: Only add an extra line break to a flag line if the next line is also a flag line.

**Example**  
```ruby
# My Class
#
# NOTE: xxxxxx xxxxx xx xxxxxxx x xxxx xx
# NOTE: yyyyyy y yyy y y yyy yyy yy yyyyy yyyy
# yyyyyy yyyyy y yyyy yyy y yyyyy yyyyyyy yy yy
# yy yyyyyyyy y yy yy yyyy yyy y yyy y yyyyyy
# NOTE: zzzzz zzzzzz zzz zzzzzz zz zz zzzz zzzzz
#
# qqqq qq q q q qqqqq qqq q qqqq qq q qqq q qq
class MyClass
end
